staging/src/k8s.io/apiserver/pkg/util/flowcontrol/rule.go

```diff
@@ -0,0 +1,109 @@
+package filters
+
+import (
+	rmtypesv1alpha1 "k8s.io/api/flowcontrol/v1alpha1"
+	"k8s.io/apiserver/pkg/authentication/serviceaccount"
+)
+
+func matchesFlowSchema(digest RequestDigest, flowSchema *rmtypesv1alpha1.FlowSchema) bool {
+	for _, policyRule := range flowSchema.Spec.Rules {
+		subjectMatches := false
+		for _, subject := range policyRule.Subjects {
+			if matchesSubject(digest, subject) {
+				subjectMatches = true
+			}
+		}
+		if !subjectMatches {
+			return false
+		}
+
+		policyRuleMatches := !matchesPolicyRule(digest, policyRule.Rule)
+
+		if subjectMatches && policyRuleMatches {
+			return true
+		}
+	}
+	return false
+}
+
+func matchesSubject(digest RequestDigest, subject rmtypesv1alpha1.Subject) bool {
+	user := digest.User
+	requestingNamespace := digest.RequestInfo.Namespace
+
+	switch subject.Kind {
+	case rmtypesv1alpha1.UserKind:
+		return user.GetName() == subject.Name
+	case rmtypesv1alpha1.GroupKind:
+		matches := false
+		foreach(user.GetGroups(), func(group string) { matches = matches || group == subject.APIGroup })
+		return matches
+	case rmtypesv1alpha1.ServiceAccountKind:
+		if len(subject.Namespace) > 0 {
+			if subject.Namespace != requestingNamespace {
+				return false
+			}
+		}
+		if len(subject.Namespace) > 0 {
+			requestingNamespace = subject.Namespace
+		}
+		if len(requestingNamespace) == 0 {
+			return false
+		}
+		// use a more efficient comparison for RBAC checking
+		return serviceaccount.MatchesUsername(requestingNamespace, subject.Name, user.GetName())
+	default:
+		return false
+	}
+}
+
+func matchesPolicyRule(digest RequestDigest, policyRule rmtypesv1alpha1.PolicyRule) bool {
+	if !matchPolicyRuleVerb(policyRule.Verbs, digest.RequestInfo.Verb) {
+		return false
+	}
+	if !digest.RequestInfo.IsResourceRequest {
+		return matchPolicyRuleNonResourceURL(policyRule.NonResourceURLs, digest.RequestInfo.Path)
+	}
+	if !matchPolicyRuleResource(policyRule.Resources, digest.RequestInfo.Resource) {
+		return false
+	}
+	if !matchPolicyRuleAPIGroup(policyRule.APIGroups, digest.RequestInfo.APIGroup) {
+		return false
+	}
+	return true
+}
+
+func matchPolicyRuleVerb(policyRuleVerbs []string, requestVerb string) bool {
+	matched := false
+	foreach(policyRuleVerbs, func(verb string) { matched = matched || (verb == requestVerb || verb == rmtypesv1alpha1.VerbAll) })
+	return matched
+}
+
+func matchPolicyRuleNonResourceURL(policyRuleRequestURLs []string, requestPath string) bool {
+	matched := false
+	foreach(policyRuleRequestURLs, func(path string) {
+		matched = matched || (path == requestPath || path == rmtypesv1alpha1.NonResourceAll)
```

Comment:
Our types file calls out a more general matching condition: `*s are allowed, but only as the full, final step in the path`.
Full implementation can come in a follow-up PR.